### ISSUE TITLE
Swift generation ignores proto packages, leading to name collisions.

### ISSUE THREAD

**Issue by mluisbrown** [square/wire#1880]
When generating Swift code, the proto package is ignored, and all files are output in the same directory. This can easily lead to name collisions where there are two types with the same name, but in different packages.

The [protobuf spec](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/descriptor.proto#L430-L434) says:

```
  // By default Swift generators will take the proto package and CamelCase it
  // replacing '.' with underscore and use that to prefix the types/symbols
  // defined. When this options is provided, they will use this value instead
  // to prefix the types/symbols defined.
  optional string swift_prefix = 39;
```

The Swift generator is not creating this CamelCase with underscores type prefix. I didn't check, but it seems likely that the Swift generator is also ignoring the `swift_prefix` file option.

I understand that Wire is [taking a different approach](https://cashapp.github.io/2020-08-19/wire-support-for-swift-part-1) with Swift but this should be opt-in. The default behaviour should be to create the type prefixes based on the proto package.

As an aside, there is absolutely *zero* documentation about the `--experimental-module-manifest` option other than the above linked blog post, which took me a long time to find.

**Comment by JakeWharton** [comment]
It is not a design goal to match what protoc does. I believe we looked at the package behavior and rejected it. I don't remember whether we discussed honoring the explicit prefix.

Swift code is designed to be generated into multiple modules using the manifest support which puts same-named types in different Swift modules allowing disambiguation by module name instead of name mangling.

**Comment by mluisbrown** [comment]
> Swift code is designed to be generated into multiple modules using the manifest support which puts same-named types in different Swift modules allowing disambiguation by module name instead of name mangling.

Absolutely. And protoc [also supports this](https://github.com/apple/swift-protobuf/blob/main/Documentation/PLUGIN.md#generation-option-protopathmodulemappings---swift-module-names-for-proto-paths), albeit in a different way to Wire, but that's fine.

However, it seems to me that the name mangling approach is defined as part of the protobuf spec, and so should be the default behaviour if no other module supporting option is specified. I'm new to protobuf, so forgive me if Google's `descriptor.proto` is not to be viewed as a hard specification.

All the above aside, since Wire seems to be (sensibly IMO) going for the module solution, it would be nice if the way to use this were properly documented, and not hidden behind an `experimental` flag.

**Comment by JakeWharton** [comment]
Yep.

I'll defer any decisions here to @efirestone and others. I'm mostly just the blunt tool that implements the design decisions.
